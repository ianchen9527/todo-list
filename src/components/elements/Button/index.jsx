import React from "react";
import styles from "./index.module.scss";
import PropTypes from "prop-types";
import cx from "classnames";

const getWrapperClassName = (disabled) =>
  cx({
    [styles["wrapper"]]: true,
    [styles["disabled"]]: disabled,
  });

const Button = ({ icon, loading, onClick, text, disabled }) => {
  return (
    <div className={getWrapperClassName(disabled)} onClick={onClick}>
      {icon && <div className={styles[`icon-${icon}`]} />}
      {text}
      {loading && <div className={styles["loading"]} />}
    </div>
  );
};

Button.propTypes = {
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  icon: PropTypes.string,
  text: PropTypes.string,
};

Button.defaultProps = {
  onClick: null,
  diabled: false,
  loading: false,
  icon: "add",
  text: "",
};

export default Button;
