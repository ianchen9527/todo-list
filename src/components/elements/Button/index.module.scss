@import "../../../stylesheets/constants/colors.scss";

.wrapper {
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  align-content: center;
  height: 40px;
  background-color: $button;
  border: 1px solid $black;
  border-radius: 5px;
  cursor: pointer;
  color: $black;
  font-family: "Roboto";
  font-size: 14px;
  line-height: 21px;

  &:hover {
    background-color: $button-hover;
  }

  &.disabled {
    opacity: 0.5;
    cursor: default;

    &:hover {
      background-color: $button;
    }
  }

  @mixin icon {
    width: 20px;
    height: 20px;
    display: block;
    background-size: 20px 20px;
    margin-right: 15px;
  }

  .icon-add {
    @include icon();
    
    background-image: url("../../../assets/images/add.svg");
  }

  .icon-edit {
    @include icon();
    
    background-image: url("../../../assets/images/edit.svg");
  }

  .loading {
    width: 20px;
    height: 20px;
    display: block;
    background-size: 20px 20px;
    background-image: url("../../../assets/images/loading.gif");
    margin-left: 15px;
  }
}
