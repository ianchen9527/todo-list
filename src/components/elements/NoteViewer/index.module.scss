@import "../../../stylesheets/constants/colors.scss";

.wrapper {
  width: 100%;
  height: 100%;
  overflow: auto;
  box-sizing: border-box;
  
  .title {
    color: $black;
    font-family: "Roboto";
    font-weight: bold;
    font-size: 20px;
    line-height: 30px;
    margin-bottom: 15px;
  }

  .loading-view {
    width: 60px;
    height: 60px;
    display: block;
    background-size: 60px 60px;
    background-image: url("../../../assets/images/loading.gif");
  }

  .content {
    border: 1px solid $black;
    padding: 15px;
    border-radius: 5px;
    box-sizing: border-box;
    margin-bottom: 15px;
    background-color: $button;
    color: $black;
    font-family: "Roboto";
    font-size: 14px;
    line-height: 21px;
    height: 50vh;
  }

  .edit-button-wrapper {
    width: 200px;
  }
}
