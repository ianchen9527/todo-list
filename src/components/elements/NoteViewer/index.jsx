import React from "react";
import styles from "./index.module.scss";
import PropTypes from "prop-types";
import Button from "../Button";
import showdown from "showdown";

const renderLoadingView = () => <div className={styles["loading-view"]} />;

const getConverter = () => new showdown.Converter();

const renderViewer = (note, onEditClick) =>
  note.id && (
    <>
      <div className={styles["title"]}>{note.title}</div>
      <div
        className={styles["content"]}
        dangerouslySetInnerHTML={{__html: getConverter().makeHtml(note.content)}}
      />
      <div className={styles["edit-button-wrapper"]}>
        <Button onClick={onEditClick} text="Edit" icon="edit" />
      </div>
    </>
  );

const NoteViewer = ({ note, loading, onEditClick }) => {
  console.log(loading);
  return (
    <div className={styles["wrapper"]}>
      {loading ? renderLoadingView() : renderViewer(note, onEditClick)}
    </div>
  );
};

NoteViewer.propTypes = {
  note: PropTypes.shape({ title: PropTypes.string, id: PropTypes.number }),
  loading: PropTypes.bool,
  onEditClick: PropTypes.func,
};

NoteViewer.defaultProps = {
  note: {},
  loading: false,
  onEditClick: null,
};

export default NoteViewer;
