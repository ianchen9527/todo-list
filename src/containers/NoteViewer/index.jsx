import { compose } from "redux";
import { connect } from "react-redux";
import {
  getNote,
  getLoading,
  getOnEditClick,
} from "../../selectors/containers/noteViewer";
import NoteViewer from "../../components/elements/NoteViewer";
import withLifecycle from "@hocs/with-lifecycle";
import { getOnDidMount } from "../../sagas/noteViewer";
import { withRouter } from "react-router-dom";

const mapStateToProps = (state, { history, match }) => ({
  note: getNote(state)(match),
  loading: getLoading(state),
  onEditClick: getOnEditClick(history)(match),
});

const mapDispatchToProps = (dispatch) => ({ dispatch });

const lifeCycle = {
  onDidMount: getOnDidMount(),
};

export default compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps),
  withLifecycle(lifeCycle)
)(NoteViewer);
