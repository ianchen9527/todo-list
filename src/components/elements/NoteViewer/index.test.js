import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import NoteViewer from "./index";
import Button from "../Button";

describe("NoteViewer", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<NoteViewer />);
  });

  describe("default props", () => {
    it("should render without crashing", () => {
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe("props.loading", () => {
    let loading;
    describe("is true", () => {
      beforeEach(() => {
        loading = true;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });

    describe("is false", () => {
      beforeEach(() => {
        loading = false;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });

      describe("props.note", () => {
        let note;
        describe("is set", () => {
          beforeEach(() => {
            note = { id: 123, title: "title", content: "content" };
            wrapper.setProps({ note });
          });

          it("should render without crashing", () => {
            expect(toJSON(wrapper)).toMatchSnapshot();
          });

          describe("props.onEditClick", () => {
            let onEditClick, event;
            beforeEach(() => {
              onEditClick = jest.fn();
              wrapper.setProps({ onEditClick });
            });

            afterEach(() => {
              onEditClick.mockClear();
            });

            describe("when .title is clicked", () => {
              beforeEach(() => {
                wrapper.find(Button).simulate("click");
              });

              it("should be triggered", () => {
                expect(onEditClick).toHaveBeenCalled();
              });
            });
          });
        });
      });
    });
  });
});
