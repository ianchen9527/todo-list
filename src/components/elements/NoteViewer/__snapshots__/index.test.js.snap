// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteViewer default props should render without crashing 1`] = `
<div
  className="wrapper"
/>
`;

exports[`NoteViewer props.loading is false props.note is set should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="title"
  >
    title
  </div>
  <div
    className="content"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>content</p>",
      }
    }
  />
  <div
    className="edit-button-wrapper"
  >
    <Button
      disabled={false}
      icon="edit"
      loading={false}
      onClick={null}
      text="Edit"
    />
  </div>
</div>
`;

exports[`NoteViewer props.loading is false should render without crashing 1`] = `
<div
  className="wrapper"
/>
`;

exports[`NoteViewer props.loading is true should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="loading-view"
  />
</div>
`;
